stm32-modules/include/common/core/queue_aggregator.hpp

```diff
@@ -0,0 +1,274 @@
+ *
+ * One limitation of this configuration is that each message queue type
+ * within MessageQueues must have a unique class definition. In the case
+ * of two message queues with the same exact message type, this may be
+ * as simple as providing a `size_t` template parameter for the queue
+ * type, and providing a different value for each "version" of the queue.
+ *
+ * @tparam MessageQueues Parameter pack of the types of all of the message
+ * queues. E.g. QueueAggregator<Queue1, Queue2, Queue3>
+ */
+template <MsgQueue... MessageQueues>
+class QueueAggregator {
+  public:
+    using TagTypes = std::variant<typename MessageQueues::Tag...>;
+    using MessageTypes = std::variant<typename MessageQueues::Message...>;
+
+    static constexpr size_t TaskCount = std::variant_size_v<TagTypes>;
+
+    static_assert(TaskCount > 0, "Must have at least one queue");
+
+    // Empty constructor
+    QueueAggregator() : _handles() {}
+
+    // Construct from a complete set of message queues. The handles are
+    // passed as a reference rather than pointers in order to simplify type
+    // deduction.
+    QueueAggregator(MessageQueues&... handles) : _handles(&handles...) {}
+
+    /**
+     * @brief Register a queue handle with the task list
+     *
+     * @tparam Idx The index of this queue
+     * @tparam Queue Type of the message queue
+     * @param queue Handle to the queue
+     * @return true if the queue handle could be registered, false
+     * if it could not be registered
+     */
+    template <MsgQueue Queue>
+    auto register_queue(Queue& queue) -> bool {
+        constexpr auto idx = get_task_idx<typename Queue::Tag>();
+        static_assert(idx < TaskCount, "Must provide a valid queue type");
+        if (check_initialized<Queue>()) {
+            // Not allowed to re-register
+            return false;
+        }
+        std::get<idx>(_handles)._handle = &queue;
+        return true;
+    }
+
+    /**
+     * @brief Get the index of a queue, as generated by the Task List
+     *
+     * @tparam Message The type of the message
+     * @return size_t A unique index for this message type
+     */
+    template <typename Queue>
+    constexpr static auto get_task_idx() -> size_t {
+        using Tag = typename Queue::Tag;
+        return TagTypes((Tag())).index();
+    }
+
+    /**
+     * @brief Send a message and automatically deduce the mailbox
+     * to forward it to.
+     *
+     * @tparam Tag The tag type of the queue to send to. This must be
+     * provided by each queue type delcared for the QueueAggregator.
+     * @tparam Message The type of message to send
+     * @param msg The message to send
+     * @param tag An instance of the tag type
+     * @return true if the message could be sent, false otherwise
+     */
+    template <typename Tag, typename Message>
+    auto send(const Message& msg, const Tag& tag = (Tag())) -> bool {
+        static_cast<void>(tag);
+        constexpr auto idx = get_task_idx<Tag>();
```

Comment:
Isn't `get_task_idx`'s tparam the `Queue` and it derives the tag from there? How does this call work?